Updating coding conventions in examples.

engineersamuel opened the pull request:
I regularly use the Office Fabric React Components, and recently used Layer.Customized.Example.tsx and Panel.Large.Example.tsx.  I am submitting this PR to update certain coding conventions in those examples.  I would like to eventually update more, however, I'd like to start small here and onboard to how you all like to do things.

(Related issue: #7907)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/7905)

engineersamuel commented on packages/office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx:
```diff
@@ -1,25 +1,22 @@
+import * as React from 'react';
 
-export interface ILayerCustomizedExampleState {
+interface IState {
```
This state is not imported anywhere so no need to be named or exported.

engineersamuel commented on packages/office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx:
```diff
@@ -1,25 +1,22 @@
-      showPanel: false,
-      trapPanel: false
-    };
-  }
```
It's not necessary to explicitly define a constructor except for very specific circumstances (like debouncing)

engineersamuel commented on packages/office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx:
```diff
@@ -1,25 +1,22 @@
-    };
-  }
+export class LayerCustomizedExample extends React.Component<{}, IState> {
+  // tslint:disable-next-line:member-access
```
State is not private in the React.Component, however the linting rules are too blunt (maybe that can't be changed) and so it wants member access to be defined for state, I think this rule is too strict here.

ecraig12345 replied:
We usually define state as public. Can you change it to public and remote the tslint:disable? (Somehow I missed this one earlier, sorry--in the future, it's probably best to leave comments that you want others to see unresolved until late in the review process.)

engineersamuel replied:
Apologies, I had fixed this and resolved, but didn't comment that I fixed.

ecraig12345 replied:
No worries, looking back at this now I realize that I forgot to look at the latest version (instead of the old one where the comment was made). This is a major reason I'm not a fan of GitHub's PR review interface... :(

engineersamuel commented on packages/office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx:
```diff
@@ -1,25 +1,22 @@
+  };
 
-  public render(): JSX.Element {
+  public render() {
```
No need to explicitly define the return type.

engineersamuel commented on packages/office-ui-fabric-react/src/components/Panel/examples/Panel.Large.Example.tsx:
```diff
@@ -1,37 +1,33 @@
-          headerText="Large Panel"
-        >
+        <DefaultButton secondaryText="Opens the Sample Panel" onClick={this._onShowPanel} text="Open Panel" />
+        <Panel isOpen={this.state.showPanel} onDismiss={this._onDismissPanel} type={PanelType.large} headerText="Large Panel">
```
Refactoring out the arrow functions, React will likely give a warning in the output for performance here if using arrow functions.

natalieethell commented on packages/office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx:
```diff
@@ -39,7 +36,7 @@ export class LayerCustomizedExample extends React.Component<{}, ILayerCustomized
           }
         >
-          {this.state.showPanel ? (
+          {this.state.showPanel && (
```
👍

natalieethell commented on packages/office-ui-fabric-react/src/components/Panel/examples/Panel.Large.Example.tsx:
```diff
@@ -1,37 +1,33 @@
+    this.setState({ showPanel: false });
+  };
+
+  private _onShowPanel = (event: React.MouseEvent<HTMLButtonElement>) => {
```
👍

natalieethell replied:
These should be `_showPanel` and `_hidePanel` right?

natalieethell replied:
Do we still not want these to be `_showPanel` and `_hidePanel` like in the other examples?

engineersamuel replied:
Great catch, just updated and pushed.  I think this illustrates how tedious function names are in general.   I'm slapping myself on the wrist right now because I was attempting to consistently name them and then myself inconsistently named then?!

natalieethell replied:
Haha no problem, thanks for updating!

natalieethell commented:
Thank you for your contributions and we're happy to hear that you regularly use Office UI Fabric React components! 

In short, we probably don't want to change things that we don't have a repo-wide convention for. If we have a mix today in conventions for things like return types and state names across those example files—which we likely do—we should probably visit those holistically. If you'd like to drive those changes holistically, you're welcome to open an issue to track and engage the wider team! Doing those changes all-up would certainly be something worth celebrating, but I'm not sure that doing it in pieces like this is best.

I think the conditional optimizations and removal of arrow functions are good, though!

engineersamuel commented:
@natalieethell Excellent, I would be very happy to drive those changes. I agree doing them piecemeal is not the best, but I suspected it would open a conversation like this.  I'll open an issue and point to this PR to get consensus.

engineersamuel commented:
@kenotron @cliffkoh @ecraig12345 Given conversations we've had from #7907 here are some points I want to cover and get verification on:
* In #7907 @kenotron mentioned making changes like this in a piecemeal fashion is preferred, so hopefully this PR will be fine in that regards
* I've tested every component I changed manually through npm run start and I updated one snapshot (very minor) and all testing passes
* Is there anything else you all would like me to do in regards to PRs like this?  I intend to make quite a few PRs as I update the examples.

Note that the CI is failing on:

```
src/components/Layer/examples/Layer.Hosted.Example.tsx:16:61 - error TS4020: 'extends' clause of exported class 'LayerHostedExample' has or is using private name 'IState'.
16 export class LayerHostedExample extends React.Component<{}, IState> {
```

which I'll work on, I wasn't aware the when generating declaration files the Props and/or State had to be exported, that still has me scratching my head a bit, I'll figure it out, but for now, would love your comments on the above points.

engineersamuel commented:
Ah, so I know why the interfaces must be exported now :) See https://github.com/Microsoft/TypeScript/issues/9865#issuecomment-234414775 Since tsconfig has declaration true and the d.ts files are generated that enforces extra restrictions.  This is unfortunate since the Examples don't need declaration files generated but the tsconfig doesn't allow includes or excludes for declarations, it's all or nothing.  I'm sure it could be worked around with a more complicated build, but not sure that is the right approach.

ecraig12345 commented on packages/office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx:
```diff
@@ -67,19 +61,19 @@ export class LayerCustomizedExample extends React.Component<{}, ILayerCustomized
   }
 
-  private _onDismissPanel = (): void => {
+  private _closePanel = () => {
```
Why this name change? At least to me it doesn't seem like a bad thing to have the callback functions be named similarly to the props they're used for.

engineersamuel replied:
Yeah names are hard, I'm aiming for consistency more than anything here.  There were some named _onDismissPanel, some with hide, some with close.  I only used `_closePanel` for consistency, but I'm not opinionated about it at all.  If anything I'd agree with you that the name should mirror the action happening, so then if `onDismissed` is the prop then options are `_onDismissed`, `_onDismissedPanel`, `_onDismissPanel`, or simply `_hidePanel`.  I'd say `_hidePanel`  probably reflects the action being taken most semantically.  Thoughts?

ecraig12345 commented on packages/office-ui-fabric-react/src/components/Panel/examples/Panel.HandleDismissTarget.Example.tsx:
```diff
@@ -52,15 +29,34 @@ export class PanelHandleDismissTargetExample extends React.Component<
   }
 
-  private _onRenderFooterContent = (): JSX.Element => {
+  private _onRenderFooterContent = () => {
```
Since this isn't a built-in, wonder if it would be better to correct the return type to React.ReactNode rather than removing it entirely?

engineersamuel replied:
This is being mapped to a built in prop though so the return type is already implicit.  If it were a completely standalone function that was not assigned to any prop I would completely agree here.  For example if this function was simply being used in the render body alone then it would be highly beneficial to constrain the type.  However, the `_onRenderFooterContent` is already constrained, you can't return the wrong thing, so then I'd see the explicit return as redundant since it adds no value.   Two examples:

```
x = () => {
    return 1;
  }

render () {
  return x();
}
```

The above works fine, but not advisable to your point, the arrow function `x` should be typed as it can return anything right now.

However with the case of `_onRenderFooterContent` the prop it is bound to requires type:

```
export interface IRenderFunction<P> {
    (props?: P, defaultRender?: (props?: P) => JSX.Element | null): JSX.Element | null;
}
```

So if I edit that `_onRenderFooterContent` to return a 1, VSCode will show an error when editing the file that the function must return type JSX.Element, so we already have the benefits of the typing implicitly, the explicit type adds no value here.

If my above arguments don't sway you I'll add the type back, it's a relatively minor thing, let me know!

ecraig12345 replied:
Thinking about this more--I'm okay with omitting the return type here since it's obvious from the method name approximately what it will return (something renderable). However, if there are cases where that's *not* obvious, I think it's best for documentation purposes to specify the return type even if it would already be constrained at compile time (by either where it's passed as a prop, or a parent class definition). And if the return type is narrower than the one defined by the prop/parent, honestly that's not the end of the world and can be corrected later if needed.

engineersamuel replied:
Sounds good, I'll echo the sentiments to being hugely in favor of strictly typing methods in general when they have no existing typed function they are being bound to.

ecraig12345 commented on packages/office-ui-fabric-react/src/components/Panel/examples/Panel.HiddenOnDismiss.Example.tsx:
```diff
@@ -29,7 +26,7 @@ export class PanelHiddenOnDismissExample extends React.Component<
   };
 
-  private _hidePanel = (): void => {
+  private _closePanel = (): void => {
```
`_hidePanel` is actually a better name in this case, since the panel is hidden not actually closed

engineersamuel replied:
Based on your previous comment to I think I agree, I'll update all `_closePanel` to `_hidePanel` as it does better reflect what is happening.

ecraig12345 replied:
That's not quite what I meant--this case does actually do something different than the others, so I think a different name would be appropriate. Usually, the panel does in fact close (disappears and all its DOM is destroyed), but in this case it's *only* hidden. (The flag was added for cases that need to maintain internal panel state between times it's displayed.)

engineersamuel replied:
Good point, changing those functions to be `_dismiss(Panel|Dialog)`.

ecraig12345 commented:
@engineersamuel Thanks for doing this! For validation, what you mentioned plus getting the component area expert to help with manual testing (probably through the PR deploy site) is a good approach. Also for examples with a codepen link, clicking it and verifying the example works in codepen.

I also think (and this is just my opinion) that even if we aren't able to update all the component examples at once to follow good practices and consistent conventions, every little bit helps. (Even better that you're willing to help drive this across the repo!) The component examples are probably the first impression that a lot of people get of Fabric, and some of them are pretty messy.

engineersamuel commented:
@ecraig12345 It's absolutely my pleasure!  I love hacking around with React/Typescript, one of my two favorite combinations of library, languages, so I'm very happy to help :)

natalieethell commented:
Nice! Looks like you need to run "rush change". Since you're only touching the example code, I think it's okay to mark the type as "none". @ecraig12345 correct me if I'm wrong.

KevinTCoughlin commented:
> Even better that you're willing to help drive this across the repo!

💯 this!

ecraig12345 commented:
@natalieethell Since example changes affect the website, it's probably best to use change type patch.

natalieethell commented on packages/office-ui-fabric-react/src/components/Panel/examples/Panel.Scroll.Example.tsx:
```diff
@@ -65,22 +63,22 @@ export class PanelScrollExample extends React.Component<
           Save
         </PrimaryButton>
-        <DefaultButton onClick={this._onClosePanel}>Cancel</DefaultButton>
+        <DefaultButton onClick={this._showPanel}>Cancel</DefaultButton>
```
Should this be `this._hidePanel`?

engineersamuel replied:
Yep, I need to do a final round of testing and verification, great catch, I'll update this.

engineersamuel replied:
Final round of testing and verification done!

engineersamuel commented on common/changes/office-ui-fabric-react/examples_update_2019-02-08-05-13.json:
```diff
@@ -0,0 +1,11 @@
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Coding conventions",
+      "type": "minor"
```
Yes definitely apologies here, I must have missed the patch option, when I ran the rush change it gave me the option for minor and major.  I'll update accordingly likely tomorrow.

ecraig12345 replied:
That's really strange... If you run into that again, let me know (it's probably something we should investigate).

engineersamuel replied:
I manually updated it to "patch".  I was not able to re-run the operation, I suspect due to temp files and something was being cached.   I don't think it is necessarily an issue with anything in this project, seemed a local fs issue, but manually updating should work as well too, let me know if everything looks good here.

ecraig12345 replied:
Manually updating to patch is fine.

natalieethell commented:
I left one more comment for Panel.Large.Example.tsx. We probably want the methods to be called `_showPanel` and `_hidePanel` like in the other examples, right?

engineersamuel commented:
@ecraig12345 Block away, I love the discussion :) I definitely am looking for consensus here, as I plan to replicate these changes to other examples, so better to well define this now than later.  I'll say upfront that I won't dig my heels in at all on this, I'll defer to whatever you and the other maintainers decide after all the points are considered.   I'll respond to each of your points and make a couple of extra ones.

> 1. It follows the convention of other components in the repo

I would suggest we change the convention in other components, at least in the examples, for reasons I'll discuss more.

> 2.  We want to model good practices in the examples, including informative interface naming

At least in the case of Component state, component state is specific to the component itself, and used only within the component, while it's possible I've never imported exported state from a component.  Therefore within a Component file, the state is internal to the Component itself, thus mirroring the name of the component name in the interface name adds no value and I find it to be redundant.   In my mind it would be akin to naming a count type number as `const countNumber: number = 0;` vs just `const count: number = 0;`  And even in that contrived example specifying number is not necessary.    With Component State, we know that the `IState` is specific to that component as it is in the same file and only used within that Component.  

Specifying the Component name in the state interface would indicate that we don't know what the state interface is for or what it encapsulates, therefore add the Component name into the state interface name so we have an indication.

And to your point, if we need a visual indication of what `IState` maps to, then I would completely agree that the Component name should be in it, but I can't find any examples where that is the case.

> 3. Makes it easier to determine via search whether the interface is being used in the repo outside that file (granted there are other ways to do this, but search is arguably the fastest/easiest)

This is definitely a valid concern; do you have any examples where exported component state is imported and used anywhere else in the repo?  


Some follow up points.

* Using `interface IState` or `type State` is both simpler and consistent across all Components.  I would advocate for this simple naming over `interface ILayerCustomizedExampleState` and `interface IPanelHandleDismissTargetExampleState` for example.  Those long names give no added value, just extra typing.  I'd argue the same logic for exporting props, `interface IProps` or `type Props` is much simpler.

* If, in the rare edge case, that exported state or props are used outside of a component, then the above naming is simple and consistent, and could easily be aliased on import.  For example if LayerCustomizedExample exported `interface IState` then one could import it like `import { IState as ILayerCustomizedExampleState } from '...';`

JasonGore commented:
Using `IProps` naming will require devs to use component namespacing or rename props on import to prevent name collision. I know you mentioned this, but consuming components and code more frequently import multiple props interfaces than not, so renaming / namespacing would be practically mandatory rather than an exception. I think this would actually increase developer hassle and not end up saving any typing. With regards to state, I don't think increased typing is really a concern outside of a given component.

We also have to consider what we would be giving up, such as searching for a specific props interface across the codebase. This is arguably the most frequent search type performed by Fabric devs. I would consider losing this to greatly outweigh any typing saved (which as I mentioned above would probably be cancelled out by renaming requirements.)

Since state interfaces also have to be exported, I think it makes sense to keep their naming convention consistent with props.

engineersamuel commented:
@JasonGore I think I see now, I did some full text searching and see a lot of Subclassed components that import state and use that state from the parent component.  Apologies everyone, I stand corrected on this point for this project!  I'll update my PR here shortly.

engineersamuel commented:
@JasonGore I'm going to slightly backtrack on my previous comment after thinking some more and say that specifically naming state and props for the core lib makes complete sense after full text searching and seeing all of the subclass components and how they import the parent component state.  However do believe that is a strong difference between the core lib style and examples in the community.  The only reason a person would need an example state/props to be named is if they intended to C&P and use it as a parent component and then created another subclass component which read the examples state, such a thing is probably not going to happen, or it'd be rare.   So in general CRUD app development, I think having the examples use a consistent name such as IState/IProps is better, but not in the core lib of this project.   All that said, I'll still defer to you all as the project maintainers, I'm just happy to be able to contribute and have this discussion :)

JasonGore commented:
I would only respond that I think the same rules that apply to us as a core lib will also apply to anybody building components on top of Fabric. Props naming collision and codebase searches would have the same level of occurrence in any but the most simplest of apps and examples. :)

engineersamuel commented:
@ecraig12345 @JasonGore Just pushed a new commit with each changed example having the state interface named as `I{ComponentName}State`

ecraig12345 commented:
Thanks for making those interface name updates! FYI there are now some merge conflicts you'll need to resolve--I happened to modify some of the same example files as part of a pass to fix ID usage.

engineersamuel commented:
@ecraig12345 Sure thing, and just resolved merge conflicts and pushed!

ecraig12345 commented:
@engineersamuel There's a snapshot update missing--can you do `npm run update-snapshots` and commit the result?

engineersamuel commented:
@ecraig12345 Good catch, actually I forgot one piece while handling the merge conflict, needed to remove an extraneous div, just did that and pushed, no need to update the snapshot as I had updated them as part of the previous changes, should be all good now.

kenotron commented:
@engineersamuel - thanks for the persistence here. This is merged!

engineersamuel commented:
@kenotron Thank you and everyone for the excellent feedback and guidance, look for more PRs very soon inline with this one!
